### PR TITLE
[SPARK-35530][ML][TESTS] Fix rounding error in DifferentiableLossAggregatorSuite with Java 11

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes an test failure of `DifferentiableLossAggregatorSuite` with Java 11.

### Why are the changes needed?

I'm personally checking whether all the tests pass with Java 11 for the current master and I found DifferentiableLossAggregatorSuite fails.
https://github.com/sarutak/spark/runs/2661859541?check_suite_focus=true#step:9:13895

The reason seems that the implementation of Blas.daxpy is different between for Java 8 and Java 11. For Java 11, `Math.fma` is used.

https://github.com/luhenry/netlib/blob/v2.2.0/blas/src/main/java/dev/ludovic/netlib/blas/Java8BLAS.java#L92
https://github.com/luhenry/netlib/blob/0053ea30b11686336cbdb8c7fceb41d59d268fa2/blas/src/main/java/dev/ludovic/netlib/blas/Java11BLAS.java#L40

To remove the rounding error, this PR changes `TestAggregator.add` with fma.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

I confirmed `DifferentiableLossAggregatorSuite` passes with both Java 8 and Java 11.

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala [hunk 1]
```diff
@@ -149,10 +152,43 @@ object DifferentiableLossAggregatorSuite {
 
     override def add(instance: Instance): TestAggregator = {
       val error = instance.label - BLAS.dot(coefficients, instance.features)
-      weightSum += instance.weight
-      lossSum += instance.weight * error * error / 2.0
-      (0 until dim).foreach { j =>
-        gradientSumArray(j) += instance.weight * error * instance.features(j)
+      val mathCls = Utils.classForName("java.lang.Math")
```
Comment (HyukjinKwon) [inline mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala:155]: @sarutak, I think we can just fix the test with tolerance e.g.) `assert(agg.weight ~== weightSum relTol 1e-5)` like the above one or https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/tuning/RandomRangesSuite.scala#L37

Up to my knowledge, this kind of precision diff is a common problem in ML side. cc @srowen, @WeichenXu123 @zhengruifeng FYI

Comment (sarutak) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala:155]: > I think we can just fix the test with tolerance at the assertion e.g.) assert(agg.weight ~== weightSum relTol 1e-5) like the above one

I considered using `~==` at first but can we accept the precision diff in this case?
The failed assertion uses `===` while some of other assertions use `~==` so I guessed `===` is intentionally used.
If we can use `~==` here, I'll fix with `~==`.

Comment (HyukjinKwon) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala:155]: I think its okay .. I did the similar things in few places before .. but should be good to have their input.

Comment (sarutak) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala]: O.K. Changed to use `~==`.

**Review by HyukjinKwon — APPROVED** [review]
Looks good up to my best knowledge.

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala (hunk 1 above)
Comment (srowen) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala]: An approximate equality test is appropriate here. I'm not sure why it varies in Java 11, but I could imagine tiny differences are possible. This is OK.

Comment (zhengruifeng) [reply mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/DifferentiableLossAggregatorSuite.scala]: > An approximate equality test is appropriate here. I'm not sure why it varies in Java 11, but I could imagine tiny differences are possible. This is OK.


+1

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**